beamer/config/commands.py

```diff
@@ -131,3 +138,255 @@ def read(
+    metavar="PATH",
+    help="Path to the keystore file.",
+)
+@click.password_option(
+    "--password",
+    type=str,
+    default="",
+    prompt=False,
+    help="The password needed to unlock the keystore file.",
+)
+@click.argument(
+    "current_state_path",
+    type=click.Path(file_okay=True, dir_okay=False, exists=True, path_type=Path),
+)
+@click.argument(
+    "desired_state_path",
+    type=click.Path(file_okay=True, dir_okay=False, exists=True, path_type=Path),
+)
+def write(
+    rpc_file: Path,
+    abi_dir: Path,
+    artifact: Path,
+    keystore_file: Path,
+    password: str,
+    current_state_path: Path,
+    desired_state_path: Path,
+) -> None:  # pylint: disable=unused-argument
+    """Read contract configuration from CURRENT_STATE_PATH and DESIRED_STATE_PATH
+    and issue transactions so that the on-chain configuration matches the desired one."""
+    beamer.util.setup_logging(log_level="DEBUG", log_json=False)
+
+    account = beamer.util.account_from_keyfile(keystore_file, password)
+    log.info("Loaded keystore file", address=account.address)
+
+    rpc_info = beamer.deploy.config.load_rpc_info(rpc_file)
+    deployment = Deployment.from_file(artifact)
+
+    assert deployment.chain is not None
+    chain_id = deployment.chain.chain_id
+    url = rpc_info[chain_id]
+    w3 = make_web3(url, account)
+    assert w3.eth.chain_id == chain_id
+    log.info("Connected to RPC", url=url)
+
+    current_config = Configuration.from_file(current_state_path)
+    desired_config = Configuration.from_file(desired_state_path, False)
+
+    _ensure_config_block_not_in_future(w3, current_config)
+    _ensure_same_tokens_have_same_addresses(current_config, desired_config)
+    _ensure_same_chain_ids(current_config, desired_config)
+
+    # Current configuration is only valid for current_config.block.
+    # This means we need to make sure that there were no config updates
+    # in block range [current_config.block + 1, latest_block].
+    start_block = BlockNumber(current_config.block + 1)
+    _ensure_no_config_updates_since(w3, deployment, start_block)
+
+    receipt = None
+    for contract, function, *args in _generate_updates(current_config, desired_config):
+        log.info("Sending transaction", call=f"{contract}.{function}({', '.join(map(str, args))})")
+        contract = deployment.obtain_contract(w3, "chain", contract)
+        call = getattr(contract.functions, function)(*args)
+        try:
+            receipt = beamer.util.transact(call)
+        except beamer.util.TransactionFailed as exc:
+            log.error("Transaction failed", exc=exc)
+            sys.exit(1)
+        else:
+            txhash = receipt.transactionHash.hex()
+            log.info("Transaction sent", block=receipt.blockNumber, txhash=txhash)
+
+    if receipt is not None:
+        # Update desired config with block number and checksum.
+        desired_config.block = receipt.blockNumber
+        desired_config.to_file(desired_state_path)
```

Comment:
From this line I can see that actually the desired state file becomes the new "historic" state file.  However is this the correct way to express that the update was successful? 
I would probably rather create a new state file updated to the current block, if that is wished (For now I would probably fine to say, update was successful and don't write anything to a file, Checks could be done afterwards with a new read).
However, a new state file output after config write is a nice to have. But I would probably not go and modify the desired config file, but rather create a proper new state file to the current block.

Reply:
Removed.